### ISSUE TITLE
Jenkins Master fails to boot after CASC Plugin upgrade to release 1.48

### ISSUE THREAD

**Issue by lenaing** [jenkinsci/configuration-as-code-plugin#1595]
Hello,

- [x] Jenkins version : 2.277.3
- [x] Plugin version : 1.48
- [x] OS : Debian Docker image (jenkins/jenkins:lts)

We rebuild our own Jenkins Master from the base Jenkins LTS image, incorporating many plugins and the configuration-as-code one.
Since the release of 1.48, our masters fails to boot without any relevant logs and just hangs at `jenkins.InitReactorRunner$1#onAttained: System config loaded`.

The only diff in our plugins list is : 
* configuration-as-code:1.48 (was 1.47)
* caffeine-api:2.9.0-17.v79e9099a51e3

When we rebuild with version 1.47, masters boot as expected.

Please find attached our CASC YAML, full plugin list and Jenkins logs.

Kind regards,

[jenkins_casc.yaml.txt](https://github.com/jenkinsci/configuration-as-code-plugin/files/6404500/jenkins_casc.yaml.txt)
[jenkins_stderr.txt](https://github.com/jenkinsci/configuration-as-code-plugin/files/6404410/jenkins_stderr.txt)
[jenkins_stdout.txt](https://github.com/jenkinsci/configuration-as-code-plugin/files/6404412/jenkins_stdout.txt)
[jenkins_plugins_list.txt](https://github.com/jenkinsci/configuration-as-code-plugin/files/6404370/jenkins_plugins_list.txt)

**Comment by timja** [comment]
@jtnord you able to take a look? ^^

**Comment by lenaing** [comment]
We had to strip down the CASC Yaml for obvious confidentiality issues, but if required I can provide a more complete one ;)

**Comment by bguerin** [comment]
Also reported here : https://issues.jenkins.io/browse/JENKINS-65518

**Comment by jtnord** [comment]
looking

**Comment by jglick** [comment]
> just hangs

Capture a thread dump. (`kill -QUIT $pid` or `jstack $pid`)

**Comment by jtnord** [comment]
The mesos cloud won't create correctly for me so I had to comment it out - but this is the config I used.
[jenkins_casc.yaml.txt](https://github.com/jenkinsci/configuration-as-code-plugin/files/6405569/jenkins_casc.yaml.txt)

**Comment by jtnord** [comment]
FTR works for me with the same file and plugins in docker also.
`docker run --env CASC_JENKINS_CONFIG=/var/jenkins_home/jenkins_casc.yaml --publish 8080:8080 --volume C:\workarea\wars\CASC_HANG_DOCKER:/var/jenkins_home jenkins/jenkins:lts`

**Comment by lenaing** [comment]
On an empty Jenkins configuration, first run is OK but second run hangs.
On an existing configuration, it just hangs.

Please find attached 3 Threads dumps on the same process.

This part seems to indicate a thread waiting for the RoleStrategy plugin, I'll try to switch back to 1.47 on this same instance to check if it starts up:
```
"ConfigurationAsCode.init" #28 daemon prio=5 os_prio=0 tid=0x00007fb630207000 nid=0x976 runnable [0x00007fb68015e000]
   java.lang.Thread.State: RUNNABLE
        at java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1934)
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.doComputeIfAbsent(BoundedLocalCache.java:2373)
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.computeIfAbsent(BoundedLocalCache.java:2356)
        at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:108)
        at com.github.benmanes.caffeine.cache.LocalLoadingCache.get(LocalLoadingCache.java:54)
        at io.jenkins.plugins.casc.impl.DefaultConfiguratorRegistry.lookup(DefaultConfiguratorRegistry.java:82)
        at io.jenkins.plugins.casc.impl.DefaultConfiguratorRegistry.internalLookup(DefaultConfiguratorRegistry.java:119)
        at io.jenkins.plugins.casc.impl.DefaultConfiguratorRegistry.access$000(DefaultConfiguratorRegistry.java:42)
        at io.jenkins.plugins.casc.impl.DefaultConfiguratorRegistry$1.load(DefaultConfiguratorRegistry.java:90)
        at io.jenkins.plugins.casc.impl.DefaultConfiguratorRegistry$1.load(DefaultConfiguratorRegistry.java:87)
        at com.github.benmanes.caffeine.cache.LocalLoadingCache.lambda$newMappingFunction$2(LocalLoadingCache.java:141)
        at com.github.benmanes.caffeine.cache.LocalLoadingCache$$Lambda$243/160341044.apply(Unknown Source)
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.lambda$doComputeIfAbsent$14(BoundedLocalCache.java:2375)
        at com.github.benmanes.caffeine.cache.BoundedLocalCache$$Lambda$296/1426841286.apply(Unknown Source)
        at java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1853)
        - locked <0x00000005c0079828> (a java.util.concurrent.ConcurrentHashMap$ReservationNode)
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.doComputeIfAbsent(BoundedLocalCache.java:2373)
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.computeIfAbsent(BoundedLocalCache.java:2356)
        at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:108)
        at com.github.benmanes.caffeine.cache.LocalLoadingCache.get(LocalLoadingCache.java:54)
        at io.jenkins.plugins.casc.impl.DefaultConfiguratorRegistry.lookupOrFail(DefaultConfiguratorRegistry.java:67)
        at io.jenkins.plugins.casc.ConfigurationContext.lookupOrFail(ConfigurationContext.java:112)
        at io.jenkins.plugins.casc.impl.configurators.DataBoundConfigurator.tryConstructor(DataBoundConfigurator.java:140)
        at io.jenkins.plugins.casc.impl.configurators.DataBoundConfigurator.instance(DataBoundConfigurator.java:76)
        at io.jenkins.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:267)
        at io.jenkins.plugins.casc.impl.configurators.DataBoundConfigurator.configure(DataBoundConfigurator.java:82)
        at org.jenkinsci.plugins.rolestrategy.casc.RoleBasedAuthorizationStrategyConfigurator.instance(RoleBasedAuthorizationStrategyConfigurator.java:58)
        at org.jenkinsci.plugins.rolestrategy.casc.RoleBasedAuthorizationStrategyConfigurator.instance(RoleBasedAuthorizationStrategyConfigurator.java:34)
        at io.jenkins.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:267)
        at io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator.lambda$doConfigure$16668e2$1(HeteroDescribableConfigurator.java:277)
        at io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator$$Lambda$320/1357238254.apply(Unknown Source)
        at io.vavr.CheckedFunction0.lambda$unchecked$52349c75$1(CheckedFunction0.java:247)
        at io.vavr.CheckedFunction0$$Lambda$301/1017878535.apply(Unknown Source)
        at io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator.doConfigure(HeteroDescribableConfigurator.java:277)
        at io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator.lambda$configure$2(HeteroDescribableConfigurator.java:86)
        at io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator$$Lambda$319/1361207555.apply(Unknown Source)
        at io.vavr.control.Option.map(Option.java:392)
        at io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator.lambda$configure$3(HeteroDescribableConfigurator.java:86)
        at io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator$$Lambda$302/1628738105.apply(Unknown Source)
        at io.vavr.Tuple2.apply(Tuple2.java:238)
        at io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator.configure(HeteroDescribableConfigurator.java:83)
        at io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator.check(HeteroDescribableConfigurator.java:92)
        at io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator.check(HeteroDescribableConfigurator.java:55)
        at io.jenkins.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:351)
        at io.jenkins.plugins.casc.BaseConfigurator.check(BaseConfigurator.java:287)
        at io.jenkins.plugins.casc.ConfigurationAsCode.lambda$checkWith$8(ConfigurationAsCode.java:750)
        at io.jenkins.plugins.casc.ConfigurationAsCode$$Lambda$280/2074231074.apply(Unknown Source)
        at io.jenkins.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:686)
        at io.jenkins.plugins.casc.ConfigurationAsCode.checkWith(ConfigurationAsCode.java:750)
        at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:735)
        at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:611)
        at io.jenkins.plugins.casc.ConfigurationAsCode.configure(ConfigurationAsCode.java:297)
        at io.jenkins.plugins.casc.ConfigurationAsCode.init(ConfigurationAsCode.java:289)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:104)
        at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:175)
        at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:296)
        at jenkins.model.Jenkins$5.runTask(Jenkins.java:1131)
        at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:214)
        at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
        at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

[jenkins_jstack1.log](https://github.com/jenkinsci/configuration-as-code-plugin/files/6405936/jenkins_jstack1.log)
[jenkins_jstack2.log](https://github.com/jenkinsci/configuration-as-code-plugin/files/6405937/jenkins_jstack2.log)
[jenkins_jstack3.log](https://github.com/jenkinsci/configuration-as-code-plugin/files/6405935/jenkins_jstack3.log)

**Comment by lenaing** [comment]
I confirm that by using 1.47 I do not have any issue with the aformentioned instance.
I'll try to remove the RoleStrategy CASC part to check if this is the blocking part, but this is not something we want to deactivate :)

**Comment by jtnord** [comment]
if it is in the `RoleBasedAuthorizationStrategyConfigurator` then it is likely dependant on your configuration which was redacted.

COuld you create a reproducable test case with https://github.com/rroemhild/docker-test-openldap as your `AuthorizationStrategy` ao that you can supply the full config unredacted?

**Comment by lenaing** [comment]
I'll try to do so.
In the meantime, please find attached the CASC Part related to the Authorization Strategy.
I just switch our users to generic ones.

[jenkins_authorization_strategy.yaml.txt](https://github.com/jenkinsci/configuration-as-code-plugin/files/6406207/jenkins_authorization_strategy.yaml.txt)

**Comment by jtnord** [comment]
#1596

**Comment by jglick** [comment]
What Java version? A fully updated 8, or something else?

**Comment by jglick** [comment]
https://github.com/ben-manes/caffeine/commits/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java has a bunch of activity after 2.9.0. Unclear to me what is at fault here.

**Comment by jglick** [comment]
Recursion is probably difficult to avoid in this case. Suggest #1596 today, then switch to using simpler “racy load” logic without a cache library if that is possible. Or it may be possible to use `HashMap.computeIfAbsent` and just keep a lock on the cache singleton—I would imagine only one thread is using this stuff anyway (should all be inside `ConfigurationAsCode.init` at least during Jenkins startup, and later CasC reload presumably does not need to run concurrently).

**Comment by ben-manes** [comment]
Note that `HashMap.computeIfAbsent` also does not support recursive computations. In JDK8 that will corrupt the hash table, while in later releases it will throw an exception.

The workaround in Caffeine is to mimic Guava, which the FAQ uses `AsyncCache` as a simple approach. Instead of computing through the map, the entry holds a future value and it is computed outside of the map. This can be done without threads with the following pseudo code.

```java
var supplier = cache.get(key, k -> Suppliers.memoize(() -> load(key)));
return supplier.get();
```

where Guava's `Supplier.memoize` uses double-checked locking to compute the value once.

If A=>B=>A then this will loop with the lock held, causing a stack overflow or other issues. In Guava it resulted in a deadlock, so I detected this case using `Thread.holdsLock`. If you go down this approach and care about that, the something like this could be done.

That said, cache recursion is tricky so you might decide it is too errorprone to support.

**Comment by jglick** [comment]
> Recursion is probably difficult to avoid in this case.

Changed my mind after actually looking at the code here. #1597 may do the trick, at the expense of sometimes doing a bit more work.

**Comment by jglick** [comment]
(and many thanks to @ben-manes for the near-real-time tech support)

**Comment by timja** [comment]
Is someone who can reproduce this able to test the hpi from https://repo.jenkins-ci.org/incrementals/io/jenkins/configuration-as-code/1.50-rc1193.ce2ce287cf96/  please?

**Comment by lenaing** [comment]
@timja I hope to have some time today to check this HPI, I'll keep you posted.
@jtnord I'll try to make a minimal test case early next week but I have to disable _a lot_ of specific setup beforehand, so I don't promise anything ;)

**Comment by ben-manes** [comment]
I updated the FAQ entry to show an example of using `Suppliers.memoize` to lazily (and atomically) compute the value outside of the cache's atomic scope. This may or may not be useful to you, but was cleaner than the async approach previously described on that wiki page. Guava's utility is a trivial implementation of double-checked locking, so that is easy to extract if you don't want the dependency. Of course doing this requires being very careful to set it up perfectly.

```java
LoadingCache<Integer, Supplier<Integer>> cache = Caffeine.newBuilder()
    .build(x -> Suppliers.memoize(() -> (x == 1) ? 1 : x * factorial(x - 1)));

int factorial(int x) {
  var memoizer = cache.get(x);
  return memoizer.get();
}
```

**Comment by lenaing** [comment]
Here are the results of my tests : 
* 1.48 ->  :x: 
* 1.49 (Current Latest) -> :heavy_check_mark:  
* 1.50-rc1193.ce2ce287cf96 -> :heavy_check_mark:

Test case to reproduce :
[plugins.txt](https://github.com/jenkinsci/configuration-as-code-plugin/files/6409601/plugins.txt)
[jenkins.yaml.txt](https://github.com/jenkinsci/configuration-as-code-plugin/files/6409603/jenkins.yaml.txt)
You'll need to change the IP to reach your LDAP test server
```
# In a shell
docker run --rm -p 10389:10389 -p 10636:10636 rroemhild/test-openldap
# In another shell
mkdir -p jenkins_home
# Initialize Jenkins and run it once to copy plugins in Jenkins Home
docker run -ti --entrypoint "/bin/bash" -p 8080:8080 -p 50000:50000 -v ${PWD}/jenkins_home:/var/jenkins_home -v ${PWD}/plugins.txt:/usr/share/jenkins/ref/plugins.txt -v ${PWD}/jenkins.yaml:/var/jenkins_home/jenkins.yaml jenkins/jenkins:lts -c "/usr/local/bin/install-plugins.sh < /usr/share/jenkins/ref/plugins.txt && echo 2 > /usr/share/jenkins/ref/jenkins.install.UpgradeWizard.state && /usr/local/bin/jenkins.sh"
# Subsequent runs that triggers the lock
docker run -ti -p 8080:8080 -p 50000:50000 -v ${PWD}/jenkins_home:/var/jenkins_home  -v ${PWD}/jenkins.yaml:/var/jenkins_home/jenkins.yaml jenkins/jenkins:lts
```
I can't trigger the lock each time however, but I get it fairly frequently with this configuration (~2 times out of 20 tries).

**Comment by jetersen** [comment]
Hopefully resolved with [v1.50](https://github.com/jenkinsci/configuration-as-code-plugin/releases/tag/configuration-as-code-1.50) :)
